File: src/main/java/dev/obscuria/tooltips/client/component/HeaderComponent.java
package dev.obscuria.tooltips.client.component;

import com.github.bsideup.jabel.Desugar;
import dev.obscuria.tooltips.client.TooltipState;
import dev.obscuria.tooltips.client.render.GuiGraphics;
import dev.obscuria.tooltips.client.tooltip.particle.GraphicUtils;
import dev.obscuria.tooltips.config.ClientConfig;
import dev.obscuria.tooltips.util.color.ARGB;
import net.minecraft.client.gui.FontRenderer;

@Desugar
public record HeaderComponent(
        TooltipState state,
        TooltipComponent title,
        TooltipComponent label,
        boolean drawSeparator,
        ARGB separatorColor
) implements TooltipComponent {
    @Override
    public int getHeight() {
        return drawSeparator ? 25 : 22;
    }

    @Override
    public int getWidth(FontRenderer font) {
        return 22 + Math.max(title.getWidth(font), label.getWidth(font));
    }

    @Override
    public void renderText(FontRenderer font, int x, int y, GuiGraphics graphics) {
        if (!ClientConfig.LABELS_ENABLED.get() || label instanceof BlankComponent) {
            title.renderText(font, 22 + x, 1 + y + 5, graphics);
        } else {
            title.renderText(font, 22 + x, 1 + y, graphics);
            label.renderText(font, 22 + x, 1 + y + title.getHeight(), graphics);
        }
    }

    @Override
    public void renderImage(FontRenderer font, int x, int y, GuiGraphics graphics) {
        state.style.slot().ifPresent(it -> it.render(graphics, x, y, 20, 20));
        state.style.effects().forEach(it -> it.renderIcon(state, graphics, x + 10, y + 10));
        state.style.icon().ifPresent(it -> it.render(state, graphics, x + 10, y + 10));

        if (!drawSeparator) return;
        final var length = getWidth(font) / 2;
        final var edgeColor = separatorColor.withAlpha(0f);
        GraphicUtils.drawHLine(graphics, x, y + 22, length, edgeColor, separatorColor);
        GraphicUtils.drawHLine(graphics, x + length, y + 22, 1 + length, separatorColor, edgeColor);
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/client/renderer/HeaderComponent.kt
package dev.obscuria.tooltips.client.renderer

import dev.obscuria.fragmentum.tools.argbOf
import dev.obscuria.tooltips.extension.GraphicsExtensions.drawHLine
import net.minecraft.client.gui.Font
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.gui.screens.inventory.tooltip.ClientTooltipComponent
import net.minecraft.client.renderer.MultiBufferSource
import org.joml.Matrix4f
import kotlin.jvm.optionals.getOrNull
import kotlin.math.max

class HeaderComponent(
    val minWidth: Int,
    val context: TooltipContext,
    val title: ClientTooltipComponent,
    val label: ClientTooltipComponent,
) : ClientTooltipComponent {

    override fun getHeight(): Int = 25

    override fun getWidth(font: Font): Int = max(minWidth, 22 + max(title.getWidth(font), label.getWidth(font)))

    override fun renderText(font: Font, x: Int, y: Int, matrix: Matrix4f, source: MultiBufferSource.BufferSource) {
        title.renderText(font, 22 + x, 1 + y, matrix, source)
        label.renderText(font, 22 + x, 1 + title.height + y, matrix, source)
    }

    override fun renderImage(font: Font, x: Int, y: Int, graphics: GuiGraphics) {
        context.style.slot.getOrNull()?.render(graphics, x, y, 20, 20)
        context.style.effects.forEach { it.renderIcon(graphics, context, x + 10, y + 10) }
        context.style.icon.getOrNull()?.render(graphics, context, x + 10, y + 10)

        val width = getWidth(font)
        graphics.drawHLine(x, y + 22, width / 2, argbOf(0x00FFFFFF), argbOf(0x60FFFFFF))
        graphics.drawHLine(x + width / 2, y + 22, 1 + width / 2, argbOf(0x60FFFFFF), argbOf(0x00FFFFFF))
    }
}